Content.Shared/RCD/Systems/RCDSystem.cs

```diff
@@ -0,0 +1,303 @@
+    [Dependency] private readonly SharedInteractionSystem _interaction = default!;
+    [Dependency] private readonly IMapManager _mapMan = default!;
+    [Dependency] private readonly INetManager _net = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly TagSystem _tag = default!;
+    [Dependency] private readonly ITileDefinitionManager _tileDefMan = default!;
+    [Dependency] private readonly IGameTiming _timing = default!;
+
+    private readonly int RCDModeCount = Enum.GetValues(typeof(RcdMode)).Length;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<RCDComponent, ExaminedEvent>(OnExamine);
+        SubscribeLocalEvent<RCDComponent, UseInHandEvent>(OnUseInHand);
+        SubscribeLocalEvent<RCDComponent, AfterInteractEvent>(OnAfterInteract);
+        SubscribeLocalEvent<RCDComponent, RCDDoAfterEvent>(OnDoAfter);
+    }
+
+    private void OnExamine(EntityUid uid, RCDComponent comp, ExaminedEvent args)
+    {
+        if (!args.IsInDetailsRange)
+            return;
+
+        var msg = Loc.GetString("rcd-component-examine-detail", ("mode", comp.Mode));
+        args.PushMarkup(msg);
+    }
+
+    private void OnUseInHand(EntityUid uid, RCDComponent comp, UseInHandEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        NextMode(uid, comp, args.User);
+        args.Handled = true;
+    }
+
+    private void OnAfterInteract(EntityUid uid, RCDComponent comp, AfterInteractEvent args)
+    {
+        if (args.Handled || !args.CanReach)
+            return;
+
+        var user = args.User;
+
+        var location = args.ClickLocation;
+        // Initial validity check
+        if (!location.IsValid(EntityManager))
+            return;
+
+        var gridIdOpt = location.GetGridUid(EntityManager);
+        if (!(gridIdOpt is EntityUid gridId) || !gridId.IsValid())
+        {
+            location = location.AlignWithClosestGridTile();
+            gridIdOpt = location.GetGridUid(EntityManager);
+            // Check if fixing it failed / get final grid ID
+            if (!(gridIdOpt is EntityUid gridId2) || !gridId2.IsValid())
+                return;
+            gridId = gridId2;
+        }
+
+        var mapGrid = _mapMan.GetGrid(gridId);
+        var tile = mapGrid.GetTileRef(location);
+        var snapPos = mapGrid.TileIndicesFor(location);
+
+        var doAfterArgs = new DoAfterArgs(user, comp.Delay, new RCDDoAfterEvent(location), uid, target: args.Target, used: uid)
+        {
+            BreakOnDamage = true,
+            NeedHand = true,
+            BreakOnHandChange = true,
+            BreakOnUserMove = true,
+            BreakOnTargetMove = args.Target != null,
+            AttemptFrequency = AttemptFrequency.EveryTick,
+            ExtraCheck = () => IsRCDStillValid(uid, comp, args.User, args.Target, mapGrid, tile, comp.Mode) // All of the sanity checks are here
```

Comment:
ExtraCheck is obsoleted.

Reply:
oh whats the new version, im pretty sure something else uses it

Reply:
actually uh the thing it says to use, "checkEvent" doesnt exist???

Reply:
I believe it's AttemptEvent.